### PR TITLE
Feature/workspace api authentication

### LINKED ISSUE #23834: Feature: Add workspace API key authentication for workspaces
### Self Checks

- [x] I have read the [Contributing Guide](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md) and [Language Policy](https://github.com/langgenius/dify/issues/1542).
- [x] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [x] I confirm that I am using English to submit this report, otherwise it will be closed.
- [x] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

I often need to manage workspaces programmatically.
If workspace operations were available via API, it would make tasks like member management and application management much easier, reducing manual work in the UI.

### 2. Additional context or comments

In the future, this functionality could be very useful when creating a CLI tool, allowing easy workspace management from the command line without relying on the UI.

### 3. Can you help us with this feature?

- [x] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by r-hashi01**
# Workspace API Authentication System Implementation

> [!IMPORTANT]
>
> 1. Make sure you have read our [contribution guidelines](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md)
> 2. Ensure there is an associated issue and you have been assigned to it
> 3. Use the correct syntax to link this PR: `Fixes #<issue number>`.

## Summary

This PR implements comprehensive workspace API authentication with scoped permissions for the Dify platform. It introduces a robust API key system that allows granular access control to workspace resources while maintaining security best practices.

### Key Features Implemented:

1. **Workspace API Key Management**
   - Full CRUD operations for workspace API keys
   - Secure token generation with RSA encryption
   - Automatic expiration handling and validation
   - Last usage tracking for security monitoring

2. **Scoped Permission System**
   - Hierarchical scope validation (admin > write > read)
   - Resource-specific permissions (workspace, apps, members)
   - Role-based access control integration
   - Comprehensive scope validation utilities

3. **Authentication Infrastructure**
   - Decorator-based authentication for workspace API endpoints
   - Database-backed permission verification
   - Secure header-based token authentication
   - Request context integration for authenticated requests

4. **API Endpoints**
   - Console endpoints for API key management (`/console/api/workspaces/current/api-keys`)
   - Workspace API endpoints with scoped authentication
   - Scope discovery endpoints for client integration
   - Token regeneration and secure key rotation

5. **Comprehensive Testing**
   - 35+ unit tests covering all authentication scenarios
   - Edge case testing for malformed requests and error conditions
   - Mock-based testing to avoid external dependencies
   - Performance testing for large scope lists

### Technical Implementation:

- **Security**: RSA-based token encryption with configurable key management
- **Database**: New `WorkspaceApiKey` model with proper indexing and relationships
- **Architecture**: Service layer pattern with clear separation of concerns
- **Validation**: Input validation with detailed error messages and type checking
- **Documentation**: Comprehensive docstrings and usage examples

This implementation provides a foundation for secure workspace-level API access while maintaining compatibility with existing Dify authentication systems.

### Files Added/Modified:

- `models/workspace_api_key.py` - New database model for API keys
- `services/workspace_api_key_service.py` - Core business logic
- `controllers/console/workspace/api_keys.py` - Console management endpoints
- `controllers/workspace/` - Workspace API endpoints with authentication
- `libs/login.py` - Authentication decorators and utilities
- `constants/workspace_scopes.py` - Scope definitions and validation
- `tests/unit_tests/` - Comprehensive test suite (35+ tests)

close #23834 
## Screenshots

| Before | After |
|--------|-------|
| No workspace API authentication system | Complete workspace API key management with scoped permissions |
| Manual workspace access control | Automated, scalable API authentication with role-based permissions |

## Checklist

- [x] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

### Additional Notes:

- **Test Coverage**: 35 comprehensive unit tests ensuring robust functionality
- **Security**: All authentication tokens are properly encrypted and validated
- **Performance**: Efficient scope checking with minimal database queries
- **Backward Compatibility**: No breaking changes to existing API endpoints
- **Error Handling**: Comprehensive error handling with informative messages
- **Code Quality**: All linting rules passed and code follows project standards

### CODE REVIEW

File: test_sql_injection.py
Comment (crazywoola) [inline test_sql_injection.py:1]: This should move into `api/`

File: api/migrations/versions/2025_08_06_2043-0fa22947e65b_merge_workspace_api_key_and_other_branch.py
Comment (laipz8200) [inline api/migrations/versions/2025_08_06_2043-0fa22947e65b_merge_workspace_api_key_and_other_branch.py:1]: Why do we have this empty migration file?

File: api/migrations/versions/2025_08_12_1818-a06820d732ce_merge_workspace_api_key_and_main_branch.py
Comment (laipz8200) [inline api/migrations/versions/2025_08_12_1818-a06820d732ce_merge_workspace_api_key_and_main_branch.py:1]: Why do we have this empty migration file?

File: web/package.json [hunk 1]
```diff
@@ -113,6 +114,7 @@
     "react-error-boundary": "^4.1.2",
     "react-headless-pagination": "^1.1.6",
     "react-hook-form": "^7.53.1",
+    "react-hot-toast": "^2.5.2",
```
Comment (crazywoola) [inline web/package.json]: We prefer not to introduce new packages unless those are not able to fulfill.

**Comment by laipz8200** [comment]
> Before you draft a PR, please create an issue to discuss the changes you want to make

**Comment by crazywoola** [comment]
Thanks for the contribution. This PR has been open for more than three months and the CI checks are currently failing, so we're closing it as stale to keep the review queue manageable. If this is still relevant, please feel free to reopen or submit a new PR with the failing checks addressed.
